plugins/resource_monitor_plugin/include/eosio/resource_monitor_plugin/file_space_handler.hpp

```diff
@@ -0,0 +1,143 @@
+
+               return true;
+            }
+         }
+
+         return false;
+      }
+
+      void add_file_system(const bfs::path& path_name) {
+         // Get detailed information of the path
+         struct stat statbuf;
+         auto status = space_provider.get_stat(path_name.string().c_str(), &statbuf);
+         EOS_ASSERT(status == 0, chain::plugin_config_exception,
+                    "Failed to run stat on ${path} with status ${status}", ("path", path_name.string())("status", status));
+
+         // If the file system containing the path is already
+         // in the filesystem list, do not add it again
+         for (auto& fs: filesystems) {
+            if (statbuf.st_dev == fs.st_dev) { // Two files belong to the same file system if their device IDs are the same.
+               ilog("${path_name}'s file system already monitored", ("path_name", path_name.string()));
+
+               return;
+            }
+         }
+
+         // For efficiency, precalculate threshold values to avoid calculating it
+         // everytime we check space usage. Since bfs::space returns
+         // available amount, we use minimum available amount as threshold. 
+         boost::system::error_code ec;
+         auto info = space_provider.get_space(path_name, ec);
+         EOS_ASSERT(!ec, chain::plugin_config_exception,
+            "Unable to get space info for ${path_name}: [code: ${ec}] ${message}",
+            ("path_name", path_name.string())
+            ("ec", ec.value())
+            ("message", ec.message()));
+
+         auto available_required = (100 - threshold) * (info.capacity / 100); // (100 - threshold) is the percentage of minimum number of available bytes the file system must maintain  
+
+         // Add to the list
+         filesystems.emplace_back(statbuf.st_dev, available_required, path_name);
+         
+         ilog("${path_name}'s file system monitored. available_required: ${available_required}, capacity: ${capacity}, threshold: ${threshold}", ("path_name", path_name.string()) ("available_required", available_required) ("capacity", info.capacity) ("threshold", threshold) );
+      }
+   
+   int space_monitor_loop() {
+      if ( is_threshold_exceeded() ) {
+         wlog("Shut down");
+         appbase::app().quit(); // This will gracefully stop Nodeos
+         return 0;
+      }
+
+      timer.expires_from_now( boost::posix_time::seconds( sleep_time_in_secs ));
+      timer.async_wait([this](auto ec) {
+         if ( ec ) {
+            wlog("Exit due to error: ${rc}, message: ${message}",
+                 ("ec", ec.value())
+                 ("message", ec.message()));
+            return 1;
+         } else {
+           // Loop over
+           space_monitor_loop();
+         }
+
+         return 2;
+      });
+
+      return 3;
```

Comment:
What do these return values mean? It doesn't look like they are ever used.

Reply:
I use those return values in unit testing so the test cases know where the function returns.